designs/2018-programmatic-style-edits/README.md

```diff
@@ -0,0 +1,401 @@
+
+At the moment, I have two other style edits in mind:
+
+1. `layout` - provides an API for tweaking spacing, semicolons, commas, and parentheses. While I have an idea about how this would work, it's a much more involved edit than `text`, so I will submit a separate RFC for this in the future.
+1. `ast` - provides an API for doing AST transforms within the context of the AST ESLint already generates. I think this could potentially be a more efficient way of doing AST transforms than using tools like Recast, but once again, this is a much more involved edit and I will submit a separate RFC for this in the future.
+
+I wouldn't want to spend the time designing these features until after style editors ships with just `text` edits first.
+
+### How will this affect the use of stylistic rules in ESLint?
+
+My hope is that once style editors are implemented, it will allow us to deprecate stylistic rules and get us out of the constant need to add options to cater to everyone's preferences about how their code should look. Instead, developers will use style editors and make the tweaks themselves, leaving us completely out of the loop.
+
+Eventually, I'd like to see all stylistic changes done inside of style editors such that no one uses stylistic rules anymore. (Admittedly, this is a long-term vision and would not happen immediately upon implementation of this proposal.)
+
+### Who will write the style edits?
+
+Once again, my hope is that this will end up being done by the ESLint community, allowing them to make their own modifications when existing style editors don't suit their purpose.
+
+My thought was that I'd spend some time writing a few basic style edits that could be included in style editors as a starting point, but I'd maintain those myself. I don't want the ESLint team to have to maintain a bunch of style edits because that situation would be too similar to what we have now with stylistic rules.
+
+### Why do style editors use an array instead of a method?
+
+An alternative design could have style editors implement an `edit()` method instead of an `edits` array. That design gives us less flexibility in a few ways:
+
+1. With individual style edits, ESLint can determine the best way to apply one style after another. For instance, with a `text` style edit, ESLint knows that it's not necessary to parse the code before passing it into the style edit. Similarly, if there are two `text` style edits in a row, ESLint knows that the result of the first style edit doesn't need to be parsed yet and can be passed directly into the second `text` style edit. This flexibility of knowing what types of style edits need to be applied is important for performance optimiziation.
+2. With a method, other style editors would have no choice but to apply all of another style editor before apply their own changes. With an array, it's possible for one style editor to pick and choose which edits from another style editor to include. 
```

Comment:
I imagine style editors' maintainers would want to add new edits in semver-minor releases, but this seems to indicate that the contents of the `edits` array within a style editor should be considered part of the editor's public API, making it immutable within a major version. This could be a case for the "Should style edits have IDs?" open question above so that style editors can compose edits from another style editor by ID rather than index.